[Clang][SVE2.1] Add intrinsics for `WHILEcc` resulting in predicate pair

momchil-velikov opened the pull request:
Add intrinsics of the form:

    svboolx2_t svwhile<cond>_b{8,16,32,64}_[{s,u}64]_x2([u]int64_t, [u]int64_t);

and their overloaded variants as specified in
https://github.com/ARM-software/acle/pull/257

momchil-velikov committed:
[Clang][SVE2.1] Add intrinsics for `WHILEcc` resulting in predicate pair

Add intrinsics of the form:

    svboolx2_t svwhile<cond>_b{8,16,32,64}_[{s,u}64]_x2([u]int64_t, [u]int64_t);

and their overloaded variants as specified in
https://github.com/ARM-software/acle/pull/257

CarolineConcatto commented on clang/include/clang/Basic/arm_sve.td:
```diff
@@ -1341,6 +1341,26 @@ def SVWHILEHS_U32 : SInst<"svwhilege_{d}[_{1}]", "Pmm", "PUcPUsPUiPUl", MergeNon
 def SVWHILEHS_U64 : SInst<"svwhilege_{d}[_{1}]", "Pnn", "PUcPUsPUiPUl", MergeNone, "aarch64_sve_whilehs", [IsOverloadWhile]>;
 }
 
+let TargetGuard = "sve2p1|sme2"  in {
```
I would suggest to remove sme2 or add the flag here and in the tests.

momchil-velikov replied:
I've added a test run with "sme2".

CarolineConcatto replied:
So that is the problem. It should not work  with sme2 when we do not have the attribute __arm_streaming_compatible . If we check the ACLE these new functions needs it:
  svboolx2_t svwhilege_b8[_s64]_x2(int64_t rn, int64_t rm)
    __arm_streaming_compatible;
I believe this is under work atm

momchil-velikov committed:
[fixup] Add a test run with "sme2"

kmclaughlin-arm commented on clang/include/clang/Basic/arm_sve.td:
```diff
@@ -1341,6 +1341,26 @@ def SVWHILEHS_U32 : SInst<"svwhilege_{d}[_{1}]", "Pmm", "PUcPUsPUiPUl", MergeNon
+  def SVWHILELO_U64_X2 : SInst<"svwhilelo_{d}[_{1}]_x2", "2nn", "PcPsPiPl",     MergeNone, "aarch64_sve_whilelo_x2">;
+  def SVWHILELS_U64_X2 : SInst<"svwhilels_{d}[_{1}]_x2", "2nn", "PcPsPiPl",     MergeNone, "aarch64_sve_whilels_x2">;
+  def SVWHILELT_U64_X2 : SInst<"svwhilelt_{d}[_{1}]_x2", "2nn", "PcPsPiPl",     MergeNone, "aarch64_sve_whilelt_x2">;
+}
```
Hi @momchil-velikov,
There were some changes made to the SME2 ACLE recently which included unifying the svwhile signed/unsigned builtins to use the same mnemonics (see https://github.com/ARM-software/acle/pull/217/commits/cdaf16f2fbe329f8951acfe11a6ae227f753d521).
I believe these changes should be applied for SVE2p1 as well?

momchil-velikov replied:
Done.

momchil-velikov committed:
[fixup] Update to the latest ACLE proposal

Remove conditions witn unsigned semantics, derive the meaning form
the operand types.

momchil-velikov committed:
[fixup] Add one more test run with sme2
